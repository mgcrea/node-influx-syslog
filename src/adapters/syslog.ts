import {FieldType} from 'influx';

export type SyslogLevel = 'emerg' | 'alert' | 'crit' | 'err' | 'warning' | 'notice' | 'info' | 'debug';

export const SEVERITY_LEVEL_VALUES: {[s in SyslogLevel]: number} = {
  emerg: 0,
  alert: 1,
  crit: 2,
  err: 3,
  warning: 4,
  notice: 5,
  info: 6,
  debug: 7
};

export const SYSLOG_SCHEMA = {
  measurement: 'syslog',
  fields: {
    version: FieldType.INTEGER,
    severity_code: FieldType.INTEGER, // eslint-disable-line @typescript-eslint/camelcase
    facility_code: FieldType.INTEGER, // eslint-disable-line @typescript-eslint/camelcase
    timestamp: FieldType.INTEGER,
    procid: FieldType.STRING,
    // msgid: FieldType.STRING,
    message: FieldType.STRING
    // sdid: FieldType.BOOLEAN
  },
  tags: ['severity', 'facility', 'host', 'hostname', 'appname']
};
