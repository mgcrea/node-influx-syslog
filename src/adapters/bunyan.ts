import {InfluxDB, ISingleHostConfig, IClusterConfig} from 'influx';
import Logger, {
  LoggerOptions as BunyanLoggerOptions,
  createLogger as createBunyanLogger,
  LogLevel as BunyanLogLevel
} from 'bunyan';
import stripAnsi from 'strip-ansi';
import {IPoint as InfluxPoint} from 'influx';
import {SyslogLevel, SEVERITY_LEVEL_VALUES, SYSLOG_SCHEMA} from './syslog';
import through2 from 'through2';
import BufferedInfluxDB from 'src/utils/BufferedInfluxDB';

export type BunyanLogObject = {
  level: BunyanLogLevel;
  msg: string;
  name: string;
  pid: string;
  time: Date;
  hostname: string;
  v?: number;
  appname?: string;
  facility?: string;
};

export const getSyslogSeverityLevel = (level: BunyanLogLevel): SyslogLevel => {
  switch (level) {
    case Logger.TRACE:
      return 'debug';
    case Logger.DEBUG:
      return 'debug';
    case Logger.INFO:
      return 'info';
    case Logger.WARN:
      return 'warning';
    case Logger.ERROR:
      return 'err';
    case Logger.FATAL:
      return 'crit';
    default:
      throw new Error(`Unhandled log level=${level}`);
  }
};

export const castSyslogInputPoint: (o: BunyanLogObject) => InfluxPoint = logObject => {
  const {time, level, name, pid, msg, hostname} = logObject;
  const {appname = 'app', facility = 'facility'} = logObject;
  const severity = getSyslogSeverityLevel(level);
  const tags = {severity, appname, facility, hostname, host: hostname};
  const timestamp = `${time.getTime()}000000`;
  const fields = {
    version: 1,
    severity_code: SEVERITY_LEVEL_VALUES[severity], // eslint-disable-line @typescript-eslint/camelcase
    facility_code: 14, // eslint-disable-line @typescript-eslint/camelcase
    timestamp,
    procid: pid,
    message: stripAnsi(msg)
  };
  return {
    measurement: 'syslog',
    tags,
    fields
  };
};

export const createSyslogStream = (influx: InfluxDB) =>
  through2.obj((chunkObj: BunyanLogObject, _enc: string, callback: () => void) => {
    const syslogPoint = castSyslogInputPoint(chunkObj);
    influx.writePoints([syslogPoint]);
    callback();
  });

export const createLogger = (
  {streams = [], ...bunyanOptions}: BunyanLoggerOptions,
  {schema = [], ...influxOptions}: ISingleHostConfig | IClusterConfig
): Logger => {
  const influx = new BufferedInfluxDB({
    schema: [...schema, SYSLOG_SCHEMA],
    ...influxOptions
  });
  return createBunyanLogger({
    streams: [
      ...streams,
      {
        stream: createSyslogStream(influx),
        level: 'trace',
        type: 'raw'
      }
    ],
    ...bunyanOptions
  });
};
