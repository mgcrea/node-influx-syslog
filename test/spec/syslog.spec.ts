import {SEVERITY_LEVEL_VALUES, SYSLOG_SCHEMA} from '../../src/adapters/syslog';

describe('syslog', () => {
  it('should properly export a schema', () => {
    expect(SEVERITY_LEVEL_VALUES).toMatchSnapshot();
  });
  it('should properly export a schema', () => {
    expect(SYSLOG_SCHEMA).toMatchSnapshot();
  });
});
