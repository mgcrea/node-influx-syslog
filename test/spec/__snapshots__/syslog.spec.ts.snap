// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syslog should properly export a schema 1`] = `
Object {
  "alert": 1,
  "crit": 2,
  "debug": 7,
  "emerg": 0,
  "err": 3,
  "info": 6,
  "notice": 5,
  "warning": 4,
}
`;

exports[`syslog should properly export a schema 2`] = `
Object {
  "fields": Object {
    "facility_code": 1,
    "message": 2,
    "procid": 2,
    "severity_code": 1,
    "timestamp": 1,
    "version": 1,
  },
  "measurement": "syslog",
  "tags": Array [
    "severity",
    "facility",
    "host",
    "hostname",
    "appname",
  ],
}
`;
